### PR TITLE
feat(kafka): meta-mapping for labels from Avro schemas on Kafka ingest

### PR DESCRIPTION
**Pull request by Deepankarkr**
Integrating this old PR: https://github.com/datahub-project/datahub/pull/7381/files as per current code.

Since Avro permits non-spec field/value pairs, a common pattern is to use this to embed business metadata. Being able to map this directly to DataHub's Owners, Tags, and Terms allows metadata to be kept closer to source, avoiding the need for Data Producers to configure transformers.

## Mapping an Avro list-of-strings to DataHub Tags

Consider the following Avro schema:
```
{
  "name": "sampleRecord",
  "type": "record",
  "tags": ["tag1", "tag2"],
  "fields": [{
    "name": "field_1",
    "type": "string",
    "tags": ["tag3", "tag4"]
  }]
}
```
The following new config allows this to be ingested as DataHub tags:
```
config:
  schema_tags_field: tags
  tag_prefix: kafka
  
```
These config values are currently the default (so they wouldn't actually be needed).

Maybe it would make more sense for the default value of `schema_tags_field` to be something like `_datahub_tags`?

Field-level tags do not automatically become Dataset-level tags.

## Meta-mapping implementation

I've re-used the meta-mapping implementation that already exists for dbt (with a small extension to support referencing nested fields).

Consider the following Avro schema:
```
{
  "name": "sampleRecord",
  "type": "record",
  "owning_team": "@Data-Science",
  "data_tier": "Bronze",
  "fields": [{
    "name": "field_1",
    "type": "string",
    "gdpr": {
      "pii": true
    }
  }]
}
```
This can now be mapped to DataHub metadata with the following config:
```
config:
  meta_mapping:
    owning_team:
      match: "^@(.*)"
      operation: "add_owner"
      config:
        owner_type: group
    data_tier:
      match: "Bronze|Silver|Gold"
      operation: "add_term"
      config:
        term: "{{ $match }}"
  field_meta_mapping:
    gdpr.pii:
      match: true
      operation: "add_tag"
      config:
        tag: "pii"
```

Note the ability to reference nested fields.

Implementation

The implementation is mostly just re-using the OperationProcessor created for the dbt source. Otherwise:

`ConfluentSchemaRegistry(KafkaSchemaRegistryBase)` to support returned multiple Aspects. The original method for backward compatibility.

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/kafka.py [hunk 1]
```diff
@@ -89,6 +89,29 @@ class KafkaSourceConfig(StatefulIngestionConfigBase, DatasetSourceConfigMixin):
         default="datahub.ingestion.source.confluent_schema_registry.ConfluentSchemaRegistry",
         description="The fully qualified implementation class(custom) that implements the KafkaSchemaRegistryBase interface.",
     )
+    schema_tags_field = pydantic.Field(
+        default="tags",
+        description="The field name in the schema metadata that contains the tags to be added to the dataset.",
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/kafka.py:94]: Noticed that this key is also checked for dataset fields level. Can we add that in description ?

File: metadata-ingestion/src/datahub/ingestion/extractor/schema_util.py [hunk 2]
```diff
@@ -317,17 +336,41 @@ def emit(self) -> Generator[SchemaField, None, None]:
                 merged_props.update(self._schema.other_props)
                 merged_props.update(schema.other_props)
 
-                tags = None
+                # Parse meta_mapping
+                meta_aspects: Dict[str, Any] = {}
+                if self._meta_mapping_processor:
+                    meta_aspects = self._meta_mapping_processor.process(merged_props)
+
+                tags = []
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/extractor/schema_util.py:344]: Adding a type here would help make this more readable and also type checks.
`tags:List[?] = []`

File: metadata-ingestion/src/datahub/ingestion/extractor/schema_util.py [hunk 3]
```diff
@@ -317,17 +336,41 @@ def emit(self) -> Generator[SchemaField, None, None]:
...
                     description = (
                         f"<span style=\"color:red\">DEPRECATED: {merged_props['deprecated']}</span>\n"
                         + description
                         if description
                         else ""
                     )
-                    tags = GlobalTagsClass(
-                        tags=[TagAssociationClass(tag="urn:li:tag:Deprecated")]
-                    )
+                    tags += [
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/extractor/schema_util.py:363]: this doesn't seem right. tags is supposedly list of strings ?

**Review by mayurinehate — CHANGES_REQUESTED** [review]
Needs quite a few changes..

1. Dataset level aspects are never emitted currently. Consider moving dataset level processing in kafka.py, similar to how description is set, if possible.
2. Tests need to be revisited for accuracy. Please update integration tests for this feature.

File: metadata-ingestion/src/datahub/ingestion/source/confluent_schema_registry.py [hunk 4]
```diff
@@ -381,6 +408,46 @@ def get_schema_metadata(
...
+                meta_aspects = self.meta_processor.process(avro_schema.other_props)
+                description = avro_schema.doc
+                for tag in avro_schema.other_props.get(
+                    self.source_config.schema_tags_field, []
+                ):
+                    all_tags.append(self.source_config.tag_prefix + tag)
+            elif schema.schema_type == "PROTOBUF":
+                pass  # not implemented
+
+        aspects: List[Any] = []
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/confluent_schema_registry.py:430]: This is the PR that adds description - https://github.com/datahub-project/datahub/pull/7974/files. the avro schema string is available in kafka.py and hence can be parsed for table level meta mapping.

File: metadata-ingestion/docs/sources/kafka/kafka.md [hunk 5]
```diff
@@ -130,3 +130,86 @@ message MessageWithMap {
   repeated Map1Entry map_1 = 1;
 }
 ```
+
+### Embedding DataHub metadata in schemas with meta mapping
```
[the remark this answers is not in the document]
Comment (Deepankarkr) [reply metadata-ingestion/docs/sources/kafka/kafka.md]: Done.

File: metadata-ingestion/src/datahub/ingestion/source/kafka.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Deepankarkr) [reply metadata-ingestion/src/datahub/ingestion/source/kafka.py:97]: Done.

File: metadata-ingestion/src/datahub/ingestion/source/kafka.py [hunk 6]
```diff
@@ -89,6 +89,29 @@ class KafkaSourceConfig(StatefulIngestionConfigBase, DatasetSourceConfigMixin):
...
+        default=True,
+        description="When enabled, applies the mappings that are defined through the meta_mapping directives.",
+    )
+    meta_mapping: Dict = pydantic.Field(
+        default={},
+        description="mapping rules that will be executed against top-level schema properties. Refer to the section below on meta automated mappings.",
+    )
+    field_meta_mapping: Dict = pydantic.Field(
+        default={},
+        description="mapping rules that will be executed against field-level schema properties. Refer to the section below on meta automated mappings.",
```
[the remark this answers is not in the document]
Comment (Deepankarkr) [reply metadata-ingestion/src/datahub/ingestion/source/kafka.py:109]: Done.

File: metadata-ingestion/src/datahub/ingestion/extractor/schema_util.py (hunk 2 above)
Comment (Deepankarkr) [reply metadata-ingestion/src/datahub/ingestion/extractor/schema_util.py]: Done.

File: metadata-ingestion/src/datahub/ingestion/source/confluent_schema_registry.py [hunk 7]
```diff
@@ -45,11 +49,14 @@ def __init__(
     ) -> None:
         self.source_config: KafkaSourceConfig = source_config
         self.report: KafkaSourceReport = report
-        self.schema_registry_client = SchemaRegistryClient(
-            {
-                "url": source_config.connection.schema_registry_url,
-                **source_config.connection.schema_registry_config,
-            }
+        # Use the fully qualified name for SchemaRegistryClient to make it mock patchable for testing.
```
[the remark this answers is not in the document]
Comment (Deepankarkr) [reply metadata-ingestion/src/datahub/ingestion/source/confluent_schema_registry.py]: Done.

File: metadata-ingestion/src/datahub/ingestion/source/kafka.py [hunk 8]
```diff
@@ -89,6 +89,29 @@ class KafkaSourceConfig(StatefulIngestionConfigBase, DatasetSourceConfigMixin):
...
+        default="tags",
+        description="The field name in the schema metadata that contains the tags to be added to the dataset.",
+    )
+    enable_meta_mapping = pydantic.Field(
+        default=True,
+        description="When enabled, applies the mappings that are defined through the meta_mapping directives.",
+    )
+    meta_mapping: Dict = pydantic.Field(
+        default={},
+        description="mapping rules that will be executed against top-level schema properties. Refer to the section below on meta automated mappings.",
```
[the remark this answers is not in the document]
Comment (Deepankarkr) [reply metadata-ingestion/src/datahub/ingestion/source/kafka.py:105]: Done.

File: metadata-ingestion/src/datahub/ingestion/extractor/schema_util.py (hunk 3 above)
Comment (Deepankarkr) [reply metadata-ingestion/src/datahub/ingestion/extractor/schema_util.py]: Done.

**Comment by hsheth2** [comment]
@Deepankarkr looks like CI is failing - please take a look

**Comment by Deepankarkr** [comment]
Closing this PR. @mayurinehate

**Pull request closed without merging**